### PR TITLE
[7.x] Refactor route caching

### PR DESCRIPTION
**Pull request by driesvints**
The past couple of weeks I've been trying to update the route caching system by using the new Symfony `CompiledUrlMatcherDumper` and `CompiledUrlMatcher` in order to get faster route resolving. Unfortunately, the benchmarks that I did so far haven't been very promising. The speed of route resolving vs the old system has almost no difference. By opening a PR here I'm hoping to see if anyone's able to chip in and help out to see if there's anything I'm missing or doing wrong. I've left some comments on the PR itself on some of the parts I'm doubting or where I believe we could do better.

Would love to get some feedback here 🙏 

Big thanks to @frankdejonge for the initial inspiration with his work on this! https://github.com/frankdejonge/framework/pull/1

### CODE REVIEW

File: src/Illuminate/Routing/Router.php [hunk 1]
```diff
@@ -1201,6 +1201,23 @@ public function setRoutes(RouteCollection $routes)
...
+     * Set the compiled route collection instance.
+     *
+     * @param  array  $routes
+     * @return void
+     */
+    public function setCompiledRoutes(array $routes)
+    {
+        foreach ($routes['attributes'] as ['methods' => $methods, 'uri' => $uri, 'action' => $action]) {
+            $this->routes->add($this->newRoute($methods, $uri, $action));
+        }
```
Comment (driesvints) [inline src/Illuminate/Routing/Router.php]: You might notice that I'm re-resolving every route here instead of setting them like beforehand. I noticed that this doesn't really makes much of a difference if I did it like before so I might be missing something here.

File: src/Illuminate/Routing/RouteCollection.php [hunk 2]
```diff
@@ -193,11 +206,27 @@ protected function matchAgainstRoutes(array $routes, $request, $includingMethod
...
+    /**
+     * Determine if a route in the compiled array matches the request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Routing\Route|null
+     */
+    protected function matchAgainstCompiledRoutes($request)
+    {
+        $context = (new RequestContext())->fromRequest($request);
+        $matcher = new CompiledUrlMatcher($this->compiledRoutes, $context);
```
Comment (driesvints) [inline src/Illuminate/Routing/RouteCollection.php]: This is the new Symfony `CompiledUrlMatcher` which should be the fastest way possible of matching routes. However, it doesn't seems to do much difference with the current system...

File: src/Illuminate/Foundation/Console/RouteCacheCommand.php [hunk 3]
```diff
@@ -58,12 +60,28 @@ public function handle()
             return $this->error("Your application doesn't have any routes.");
         }
 
+        $symfonyRoutes = new SymfonyRouteCollection();
+        $attributes = [];
+
         foreach ($routes as $route) {
             $route->prepareForSerialization();
+
+            $name = $route->getName() ?? Str::random();
```
Comment (driesvints) [inline src/Illuminate/Foundation/Console/RouteCacheCommand.php]: For the new system I needed every route to have a name so if a route doesn't has a name, I generate one for it. I don't know if this has complications for any existing app. Does anyone have any thoughts if this could affect some apps?

**Comment by 36864** [comment]
Define "almost no difference" and what % you're looking for, and in what circumstances.

Do you have any benchmark code for people to test this out and try to improve?

File: src/Illuminate/Foundation/Console/RouteCacheCommand.php [hunk 4]
```diff
@@ -58,12 +61,28 @@ public function handle()
...
+        foreach ($routes as &$route) {
             $route->prepareForSerialization();
+
+            // If the route doesn't have a name, we'll generate one for it
+            // and re-add the route to the collection. This way we can
+            // add the route to the Symfony route collection.
+            if (! $name = $route->getName()) {
+                $route->name($name = Str::random());
+
+                $routes->add($route);
```
Comment (driesvints) [inline src/Illuminate/Foundation/Console/RouteCacheCommand.php]: For the new system I needed every route to have a name so if a route doesn't has a name, I generate one for it. I don't know if this has complications for any existing app. Does anyone have any thoughts if this could affect some apps?

File: src/Illuminate/Routing/RouteCollection.php [hunk 5]
```diff
@@ -356,4 +399,41 @@ public function count()
...
+        $symfonyRoutes = new SymfonyRouteCollection();
+
+        foreach ($this->getRoutes() as $route) {
+            // If the route doesn't have a name, we'll generate one for it
+            // and re-add the route to the collection. This way we can
+            // add the route to the Symfony route collection.
+            if (! $name = $route->getName()) {
+                $route->name($name = Str::random());
+
+                $this->add($route);
```
Comment (driesvints) [inline src/Illuminate/Routing/RouteCollection.php]: For the new system I needed every route to have a name so if a route doesn't has a name, I generate one for it. I don't know if this has complications for any existing app. Does anyone have any thoughts if this could affect some apps?

**Comment by driesvints** [comment]
@36864 definitely. I used the `ab` benchmark tool to do my benchmarks. 

Here's a benchmark for 800 dynamic routes:

**`routes.php`:**

```php
<?php

Route::get('/', 'Controller@index');

for ($i = 0; $i < 800; $i++) {
    Route::get("/foo/{bar}/$i", 'Controller@index')->name("non-static-$i");
}
```

Benchmark for **current cached routing** returns 269 requests per second:

<details>
  <summary>View benchmark</summary>

```bash
$ ab -t 10 -c 10 http://test-routes.test/foo/wrwer/799
This is ApacheBench, Version 2.3 <$Revision: 1843412 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking test-routes.test (be patient)
Finished 2691 requests


Server Software:        nginx/1.17.7
Server Hostname:        test-routes.test
Server Port:            80

Document Path:          /foo/wrwer/799
Document Length:        2426 bytes

Concurrency Level:      10
Time taken for tests:   10.001 seconds
Complete requests:      2691
Failed requests:        0
Total transferred:      9127040 bytes
HTML transferred:       6528366 bytes
Requests per second:    269.08 [#/sec] (mean)
Time per request:       37.164 [ms] (mean)
Time per request:       3.716 [ms] (mean, across all concurrent requests)
Transfer rate:          891.23 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:    19   37  34.9     33    1582
Waiting:       19   37  34.9     33    1581
Total:         19   37  34.9     33    1582

Percentage of the requests served within a certain time (ms)
  50%     33
  66%     38
  75%     41
  80%     43
  90%     50
  95%     55
  98%     65
  99%     72
 100%   1582 (longest request)
```
</details>

Benchmark for **new cached routing** returns 221 requests per second:

<details>
  <summary>View benchmark</summary>

```bash
$ ab -t 10 -c 10 http://test-routes-caching.test/foo/ewrrre/799                ~/Sites/test-routes-caching
This is ApacheBench, Version 2.3 <$Revision: 1843412 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking test-routes-caching.test (be patient)
Finished 2215 requests


Server Software:        nginx/1.17.7
Server Hostname:        test-routes-caching.test
Server Port:            80

Document Path:          /foo/ewrrre/799
Document Length:        2426 bytes

Concurrency Level:      10
Time taken for tests:   10.003 seconds
Complete requests:      2215
Failed requests:        0
Total transferred:      7512576 bytes
HTML transferred:       5373590 bytes
Requests per second:    221.44 [#/sec] (mean)
Time per request:       45.160 [ms] (mean)
Time per request:       4.516 [ms] (mean, across all concurrent requests)
Transfer rate:          733.44 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:    12   45  99.1     32    2512
Waiting:       12   45  99.1     32    2512
Total:         12   45  99.1     32    2513

Percentage of the requests served within a certain time (ms)
  50%     32
  66%     35
  75%     38
  80%     39
  90%     50
  95%     85
  98%    244
  99%    280
 100%   2513 (longest request)
  90%     50
  95%     55
  98%     65
  99%     72
 100%   1582 (longest request)
```
</details>

What must be noted is that I get widely different results with `ab` from time to time for some reason. I have no idea why that is.

**Comment by 36864** [comment]
Thanks. I'll see if I can dig into this tonight.

Looks like the new system is slower on average, but has a slightly faster best-case.

It might be worth benchmarking cases with different number of routes to determine if the changes made have an impact on every app or only those with many routes (although I guess if an app only has 5 routes no amount of fiddling with the caching mechanism would have a significant impact)

As for the random variance you're getting, try reducing the concurrency level for `ab`.

File: src/Illuminate/Routing/RouteCollection.php (hunk 5 above)
Comment (frankdejonge) [reply src/Illuminate/Routing/RouteCollection.php]: Isn't there a name you can create based on the input? For example, the controller::handler? or for callbacks a callback_[index]?

Comment (driesvints) [reply src/Illuminate/Routing/RouteCollection.php]: Callbacks aren't supported for route caching but I could look into the controller::handler syntax yeah 👍

Comment (36864) [reply src/Illuminate/Routing/RouteCollection.php]: There is a risk of conflicting with a name that's already defined, even if it's a small risk.

Maybe have a clear prefix, such as "unnamed.{controller}@{method}"?

**Comment by danilopinotti** [comment]
Did you try to use json_encode/json_decode instead of serialize/unserialize?

File: src/Illuminate/Foundation/Console/RouteCacheCommand.php [hunk 6]
```diff
@@ -98,12 +101,16 @@ protected function getFreshApplication()
...
      * @return string
      */
-    protected function buildRouteCacheFile(RouteCollection $routes)
+    protected function buildRouteCacheFile(RouteCollection $routes, array $compiledRoutes)
     {
         $stub = $this->files->get(__DIR__.'/stubs/routes.stub');
 
-        return str_replace('{{routes}}', base64_encode(serialize($routes)), $stub);
+        $replaced = str_replace('{{routes}}', base64_encode(serialize($routes)), $stub);
+        $replaced = str_replace('{{compiledRoutes}}', base64_encode(serialize($compiledRoutes)), $replaced);
```
Comment (nicolas-grekas) [inline src/Illuminate/Foundation/Console/RouteCacheCommand.php]: Is it possible to use raw php instead of base64+serialize?
I'd guess the cost of decoding+unserialing completely dominates the matching time, and this is why things are not faster?

**Comment by donnysim** [comment]
Don't know, for me the new version is faster:

Windows 10, PHP 7.4.1, Apache
`ab -t 10 -c 10`

<details>
  <summary>Laravel 6</summary>

```
This is ApacheBench, Version 2.3 <$Revision: 1826891 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking laravel6.test (be patient)
Finished 2529 requests


Server Software:        Apache/2.4.35
Server Hostname:        laravel6.test
Server Port:            80

Document Path:          /foo/wrwer/799
Document Length:        2 bytes

Concurrency Level:      10
Time taken for tests:   10.005 seconds
Complete requests:      2529
Failed requests:        0
Total transferred:      2452280 bytes
HTML transferred:       5058 bytes
Requests per second:    252.77 [#/sec] (mean)
Time per request:       39.562 [ms] (mean)
Time per request:       3.956 [ms] (mean, across all concurrent requests)
Transfer rate:          239.36 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.5      0       1
Processing:    15   37  97.8     23     934
Waiting:       15   37  97.8     22     933
Total:         15   38  97.8     23     934

Percentage of the requests served within a certain time (ms)
  50%     23
  66%     24
  75%     24
  80%     25
  90%     26
  95%     28
  98%    518
  99%    682
 100%    934 (longest request)
```
</details>


<details>
  <summary>Laravel 7 refactor-route-caching</summary>

```
This is ApacheBench, Version 2.3 <$Revision: 1826891 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking laravel.test (be patient)
Finished 3189 requests


Server Software:        Apache/2.4.35
Server Hostname:        laravel.test
Server Port:            80

Document Path:          /foo/wrwer/799
Document Length:        2 bytes

Concurrency Level:      10
Time taken for tests:   10.010 seconds
Complete requests:      3189
Failed requests:        0
Total transferred:      3182876 bytes
HTML transferred:       6380 bytes
Requests per second:    318.57 [#/sec] (mean)
Time per request:       31.390 [ms] (mean)
Time per request:       3.139 [ms] (mean, across all concurrent requests)
Transfer rate:          310.51 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.5      0       1
Processing:    15   30  73.0     21     816
Waiting:       13   30  73.0     21     816
Total:         15   31  73.0     21     817

Percentage of the requests served within a certain time (ms)
  50%     21
  66%     22
  75%     23
  80%     23
  90%     24
  95%     25
  98%     28
  99%    548
 100%    817 (longest request)
```
</details>

**Comment by driesvints** [comment]
> Did you try to use json_encode/json_decode instead of serialize/unserialize?

No. I don't think that'll be any faster though?

> Don't know, for me the new version is faster:

Thanks for testing! Did you try running the benchmarks a couple of times to see if you get similar results?

File: src/Illuminate/Routing/RouteCollection.php (hunk 5 above)
Comment (driesvints) [reply src/Illuminate/Routing/RouteCollection.php]: > Maybe have a clear prefix, such as "unnamed.{controller}@{method}"?

Yeah we can definitely do that 👍

**Comment by donnysim** [comment]
Yeah, around 10 times each, with opcache and without, new version was always faster by 30+ requests/s.

File: src/Illuminate/Foundation/Console/RouteCacheCommand.php (hunk 6 above)
Comment (driesvints) [reply src/Illuminate/Foundation/Console/RouteCacheCommand.php]: @nicolas-grekas I actually also tested with the dumped PHP array file from the `CompiledUrlMatcherDumper` but that didn't lead to better results. 

On that specific case: I had to use reflection to access the `generateCompiledRoutes` method so I could combine the content with the compiled routes from Laravel. I could create two separate cache files but I'd like to keep everything together in one if possible. Any chance we can make the `generateCompiledRoutes` method public or extract the contents to a new public method?

**Comment by danilopinotti** [comment]
> No. I don't think that'll be any faster though?

In some cases, json_encode/json_decode is faster than serialize/unserialize.
In a very simple test, these were the results:
```
$data = (object)['a' => 'b'];
$jsonData = json_encode($data)
$serializedData = serialize($data)

stress(function () use ($jsonData) {json_decode($jsonData);}, 1000000, false) // 512ms
stress(function () use ($serializedData) {unserialize($serializedData);}, 1000000, false) // 1242ms
```

The stress method only executes the callback in loop and returns the time in miliseconds.

This was a very simple test, the results may not be similar for routes.

File: src/Illuminate/Routing/RouteCollection.php (hunk 5 above)
Comment (driesvints) [reply src/Illuminate/Routing/RouteCollection.php]: @frankdejonge I decided to keep the randome string since it's concise and random. Otherwise you get the action twice when you do `route:list`:

Screenshot 2020-01-21 at 21 30 18

**Comment by danilopinotti** [comment]
> No. I don't think that'll be any faster though?

In very simple tests, json_decode is faster, but, with complex and larger objects, json_decode is more slow than unserialize.

Really, will not bring improvements.

**Comment by derekmd** [comment]
> > Did you try to use json_encode/json_decode instead of serialize/unserialize?
> 
> No. I don't think that'll be any faster though?
> 

`igbinary_serialize()` ( https://github.com/igbinary/igbinary ) can be much faster than `serialize()` because it generates a much smaller string. e.g., for 838 Laravel routes, the stored bootstrap/cache/routes.php file was reduced from 4.29mb to 0.88mb.

Only problem is rarely is this extension available out-of-the-box with most PHP hosting. `pecl install igbinary`, `yum install php73-pecl-igbinary`, etc. is usually required in each environment to make this caching solution work.

**Comment by nicolas-grekas** [comment]
Sorry for the noob question, but could anyone provide my with a small app that would run this PR and allow me to run the `ab` benchs too?

**Comment by gocanto** [comment]
> Sorry for the noob question, but could anyone provide me with a small app that would run this PR and allow me to run the ab benchs too?

@nicolas-grekas 

you will have to create a dummy [app](https://laravel.com/docs/6.x) and then pull this version thru [composer](https://github.com/laravel/laravel/blob/master/composer.json#L13) when requiring the `framework`

or you could clone this [repo](https://github.com/laravel/laravel) and do the same ^

**Comment by nicolas-grekas** [comment]
@gocanto thanks, that's the parts I figured out, but then, I don't know what to do, which files to patch, how to compile the routes, etc :)

**Comment by 36864** [comment]
>I don't know what to do, which files to patch, how to compile the routes, etc :)

Once you have this branch pulled, you'll need to empty the contents of `routes/api.php`, add a method named `index` to the `App\Http\Controllers\Controller` class, and then add the test routes to `routes/web.php`.

Once you've done that, run `php artisan route:cache` to cache the routes.

**Comment by gocanto** [comment]
@nicolas-grekas I see. 

If you want to compile the `routes`, you will have to go into your dummy app console line and type `php artisan route:cache`.

After this is done, you should be able to see a compiled routes file within your `boostrap/cache` directory.

**Comment by 36864** [comment]
After typing out all the steps I realized it would have been easier to just upload my test project. Here you go: https://github.com/36864/laravel-route-cache-test

Just clone that, run `composer install`, and then `php artisan route:cache`.

**Comment by nicolas-grekas** [comment]
Thanks @36864 that helped a lot!

So here are my quick findings:

The biggest perf hog is the call to `app('router')->setRoutes()` - actually it's the unserialization+decoding of its argument - a huge payload.

With the current state of the PR, I'm at 74req/s

If I comment the call to `setRoutes()` in the routes cache, I get 180req/s.

Then, if I switch to using `var_export()` instead of base64+serialize, I top at 250 req/s.

My reco would be to use `var_export()` for the call to `setCompiledRoutes()`: it's been optimized for leveraging opcache static arrays.

For `setRoutes()`, using b64+serialize looks fine, but I would recommend finding a way to NOT do the call - it's so costly to create this RouteCollection. At least, it should be done lazily, if possible.

**Comment by danilopinotti** [comment]
Another problem that I found in current route cache implementation is the high RAM consume in a request when system has a lot of routes (1000+). It increments about 10MB of RAM by each request. Currently, I needed to disable route cache when production because this problem.

Will this problem be solved?

**Comment by driesvints** [comment]
@nicolas-grekas thank you for helping out here! Some thoughts on your feedback:

> If I comment the call to setRoutes() in the routes cache, I get 180 req/s.

Hmm, technically it's not possible that the current PR will work without the call. If I comment it out myself I get a 404 when I visit any route but I do get improved results from the `ab` testing tool and it's not showing any failures (which is bizar because the requests return 404's).

> Then, if I switch to using var_export() instead of base64+serialize, I top at 250 req/s.

If I use var_export for the regular routes I get the following error:

```
Call to undefined method Illuminate\Routing\RouteCollection::__set_state()
```

I guess you implemented this method on the RouteCollection?

> My reco would be to use var_export() for the call to setCompiledRoutes(): it's been optimized for leveraging opcache static arrays.

I've just pushed a commit with switching to var_export for `setCompiledRoutes`. Already seeing some better results indeed!

> For setRoutes(), using b64+serialize looks fine, but I would recommend finding a way to NOT do the call - it's so costly to create this RouteCollection. At least, it should be done lazily, if possible.

I'll try to look into this tomorrow if that's possible.

One more question: what exact benchmark command are you using? Thanks again for helping out!

**Comment by driesvints** [comment]
@taylorotwell let me know that with the latest version he isn't seeing improvements himself.

**Comment by nicolas-grekas** [comment]
> Call to undefined method Illuminate\Routing\RouteCollection::__set_state()

Actually I didn't, I just commented out the call :)

But that's definitely the next thing to optimize. Could be by implementing this method, or by creating a new `CompiledRouteCollection` that would take a simple array as argument (the class would create individual route objects on demand.)

> what exact benchmark command are you using?

It was `ab -n500` with `php artisan serve` :)

**Comment by GrahamCampbell** [comment]
Was opcache enabled?

**Comment by GrahamCampbell** [comment]
I'm assuming opcache would execute the base64 decode ahead of time, if enabled, which could explain why you and Dries were seeing some differences.

**Comment by nicolas-grekas** [comment]
It was enabled for me, but it would be totally new knowledge to me if opcache were to resolve b64 ahead of time. What do you base this idea on?

**Comment by GrahamCampbell** [comment]
Opcache already does constant folding and some ahead of time evaluations, like is_string on a value that is known to be a string. I wouldn't be surprised if it did the base64 decode a head of time, but I've not checked the source. If it doesn't... maybe it should. ;)

**Comment by nicolas-grekas** [comment]
For reference, here is what I know on the topic, with links to the source: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048
TL;DR, only a few carefully selected functions benefit from compile-time optimizations and adding more won't happen AFAIK because it [would slow down the engine](https://github.com/php/php-src/pull/3360#pullrequestreview-148337937).

**Comment by driesvints** [comment]
I also have opcache enabled.

> Could be by implementing this method

Hmm, since `CompiledRoute` doesn't implements this, this isn't really an option I think 🤔 

> or by creating a new CompiledRouteCollection that would take a simple array as argument

Hmm this was the path I took first but it overcomplicated the router (because it needs to have the `$routes` property set). I'm a bit hesitant on trying that again.

I really wonder why there's so much difference in the test results between yours and mine. With your `ab` test from above I get 85.74 rps on the current route cache and 93.85 rps on the current implementation (with opcache enabled). I'm a bit at a loss and don't really know what to do any further.

**Comment by driesvints** [comment]
I think I finally managed to crack it. I ditched the setRoutes call like you suggested @nicolas-grekas and I'm not finally seeing much better results. I still need to find a way to lazy-load the routes when `getRoutes` is called upon the collection but I think we're finally getting somewhere. 

I'll pick this up on monday.

**Comment by frankdejonge** [comment]
@driesvints `/s/not finally/now finally` by any chance? The rest of the message seems very promising!

**Comment by frankdejonge** [comment]
@driesvints I have an idea to allow closures to be included in the route caching as well, we should skype some time soon :P

**Comment by driesvints** [comment]
Yes I meant now! 😶

Sweet! Let's chat soon :)

**Comment by GrahamCampbell** [comment]
> I have an idea to allow closures to be included in the route caching as well, we should skype some time soon :P

We did used to support closure serialising at one point, but people kept doing stuff that implicitly "used" the app service container, like by passing through models, which made a mess they couldn't understand.

**Comment by frankdejonge** [comment]
@GrahamCampbell I had a different approach in mind than using serialization for that ;-)

File: src/Illuminate/Routing/RouteCollection.php [hunk 7]
```diff
@@ -4,6 +4,7 @@
 
 use ArrayIterator;
 use Countable;
+use Illuminate\Container\Container;
```
Comment (X-Coder264) [inline src/Illuminate/Routing/RouteCollection.php]: Shouldn't this reference the container contract (along with all the docblocks as well)?

Comment (driesvints) [reply src/Illuminate/Routing/RouteCollection.php]: Yeah true thanks. I'm refactoring a lot here btw but will take that into account.

Comment (X-Coder264) [reply src/Illuminate/Routing/RouteCollection.php]: 👍

Comment (driesvints) [reply src/Illuminate/Routing/RouteCollection.php]: Just checked and I unfortunately need to pass the actual container here because the Route atm requires this. I'm thinking of refactoring that as well but that'll be for a different pr.

**Comment by driesvints** [comment]
I've pushed some more changes. This is pretty much the version I intend to send in. Seeing very promising results here (165rps for the new caching vs 85rps for the old way with `ab -n 500`).

@nicolas-grekas if you could also give these changes another go that'd be great. Thanks for all your help so far!

File: src/Illuminate/Routing/AbstractRouteCollection.php [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            }
+        }
+
+        return $others;
+    }
+
+    /**
+     * Determine if a route in the array matches the request.
+     *
+     * @param  array  $routes
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/AbstractRouteCollection.php]: more specific doc here?

File: src/Illuminate/Routing/AbstractRouteCollection.php [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+     * @param  bool  $includingMethod
+     * @return \Illuminate\Routing\Route|null
+     */
+    protected function matchAgainstRoutes(array $routes, $request, $includingMethod = true)
+    {
+        [$fallbacks, $routes] = collect($routes)->partition(function ($route) {
+            return $route->isFallback;
+        });
+
+        return $routes->merge($fallbacks)->first(function (Route $route) use ($request, $includingMethod) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/AbstractRouteCollection.php]: do we want the typechck here, or should we remove it like the rest of the code?

File: src/Illuminate/Routing/AbstractRouteCollection.php [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+            return $route->matches($request, $includingMethod);
+        });
+    }
+
+    /**
+     * Get a route (if necessary) that responds when other available methods are present.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  array  $methods
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/AbstractRouteCollection.php]: `string[]`

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Routing/AbstractRouteCollection.php (hunk 9 above)
Comment (driesvints) [reply src/Illuminate/Routing/AbstractRouteCollection.php]: This part hasn't been touched so I'm gonna leave it be.

**Comment by nicolas-grekas** [comment]
I won't be able to check out soon, but this looks good so far!

**Comment by taylorotwell** [comment]
@driesvints there is some broken behavior here when using `Route::fallback`. (https://laravel.com/docs/6.x/routing#fallback-routes). You can see that the normal Route collection has some specific partitioning happening around these routes whereas cached routes don't.

You can see the bug if you define a route setup like this (the domain part is irrelevant as I was testing that as well):

```php
Route::domain('foo.laravel.test')->group(function () {
    Route::fallback(FooFallbackController::class);
    Route::get('/', FooDomainTestController::class);
});

Route::domain('bar.laravel.test')->group(function () {
    Route::fallback(BarFallbackController::class);
    Route::get('/', BarDomainTestController::class);
});
```

Without route caching, the fallback response is correctly only returned when no other routes match... with route caching enabled, the fallback route's response will over-shadow the `/` route and will be returned when you hit that route (as well as any other route).

**Comment by taylorotwell** [comment]
It is likely worth adding tests for CompiledRouteCollection and its various methods as well as RouteCollection (if we don't have tests for that class already).

**Comment by garygreen** [comment]
@driesvints 

> I've pushed some more changes. This is pretty much the version I intend to send in. Seeing very promising results here (165rps for the new caching vs 85rps for the old way with ab -n 500).

Can you clarify what the speed is WITHOUT any caching at all? From previous testing in our application when caching is enabled it actually has a negative effect due to the large deserialization of the route payload - I'm sure this PR is an improvement over that but is it that much quicker than no caching at all? Registering 1000 routes is actually pretty quick in Laravel without any caching.

**Comment by taylorotwell** [comment]
@garygreen I get 145 requests per second totally uncached and 517 requests per second cached on an application with 800 GET routes.

**Comment by garygreen** [comment]
Just done some testing with this in our real world app with around 400 routes and I'm actually seeing a negative impact on speed with this PR. There also appears to be a slight regression in speed between Laravel 6 and 7 even without caching.

Laravel 6 (without caching): 130 rq/s
Laravel 7 (without caching): 105 rq/s
Laravel 7 (with caching): 85 rq/s

Very odd. I've double checked that it's definitely using the new caching file on local by adding a random debug statement in routes.php and making sure it doesn't fire.

@taylorotwell how did you manage to go from 145 requests to 517 requests a second? Would it be worth sharing a fork of your setup?

**Comment by taylorotwell** [comment]
No I'm not preloading anything. Both @driesvints and I are seeing the improvements. I just took this branch and did this in a routes file:

```php
for ($i = 0; $i < 800; $i++) {
    Route::get('/test-'.$i.'/{foo}', 'TestController');
}
```

```php
class TestController extends Controller
{
    public function __invoke($foo)
    {
        return 'Invoked - '.$foo;
    }
}
```

Then `php artisan route:cache` and `ab -t 10 -c 1 http://laravel.test/test-799/foo` on the command line.

**Comment by garygreen** [comment]
@taylorotwell thanks for the information. Just tested using your benchmark and it does indeed speedup with caching:

22 rq/s without caching
58 rq/s with caching

One thing I have noticed is in the compiled caching file there is a big block of what looks like optimized regex:

image

However in our apps compiled route caching file there doesn't appear to be any of these kind of optimizations:

image

I'm wondering if the repetitive nature of `test-X` is allowing Symphony to do optimizations on those routes and that's why in your benchmark it's quicker - however in our real world app such optimizations haven't been applied so we're not seeing any speed difference.

**Comment by garygreen** [comment]
What's odd as well is in our app we do have repetitive urls that start with "/profile/some-action" "/profile/another-action" so I would of thought we would get some kind of optimized routing on those.

**Comment by nicolas-grekas** [comment]
@garygreen you might want to read https://medium.com/@nicolas.grekas/making-symfonys-router-77-7x-faster-1-2-958e3754f0e1 and https://medium.com/@nicolas.grekas/making-symfony-router-lightning-fast-2-2-19281dcd245b to know how this works.

**Comment by garygreen** [comment]
@nicolas-grekas interesting links. I can't claim to understand all the implementation details of how Symphony is optimizing routes but it does seem like in our application it's not optimizing any of the routes even though according to those articles it should be generating a optimized regex for repeating patterns/prefixes.

Wondering if maybe this PR hasn't called Symphony api correctly to optimize the routes?

**Comment by nicolas-grekas** [comment]
@garygreen:
>22 rq/s without caching
>58 rq/s with caching

That's a huge improvement. I don't know what makes you think routes are not optimized in your case, but that's just not possible as all routes are. The screenshot you posted shows some static routes. The regexp is used only for dynamic routes (the one which contains placeholders). If all/most your routes are static, they were already quite optimized and the diff won't but as big as with dynamic routes. That's maybe why you're seeing a lower diff. But be assured, all works as expected, this is battle tested :)

**Comment by driesvints** [comment]
Verified that the fallback issue isn't present in this pr. Think except for tests that we're pretty much there then.

@nicolas-grekas I'm wondering if it makes a difference if we just returned an array from the compiled routes file and do the setCompiledRoutes call in the service provider instead. Not sure if that makes a difference for opcache? 

https://github.com/laravel/framework/pull/31188/files#diff-285965dba4ad1f52b822bbae17e94ca6R14

**Comment by nicolas-grekas** [comment]
Not preference in terms of opcache: both ways would provide equal opcache-ability. As fits best for you.

**Comment by driesvints** [comment]
Thank you ALL for helping getting this merged in! Especially @nicolas-grekas and @frankdejonge 👏👏

**Comment by yaquawa** [comment]
Hi @driesvints 
Thanks for the efforts of this!

But, now you can't set your own validators....
Before this PR, you can replace the validators so we can control the validation rules of routes.

```php
Route::$validators = [
    new MethodValidator,
    new SchemeValidator,
    new HostValidator,
    new UriValidator,
];
```

This PR completely breaks my code if I cached the routes.

Can we have a change to choose one of them?

**Comment by driesvints** [comment]
@yaquawa this was never documented or publicly supported sorry. You can technically still use that (although I don't recommend to) if you don't use route caching.

**Comment by yaquawa** [comment]
Hi @driesvints, but I don't think it's a good idea to change the behavior of how router works just because caching is applied or not. Why not just use one of them? Two implementations works together feels so weird.

**Comment by driesvints** [comment]
@yaquawa sorry we have no plans to change this.

**Comment by yaquawa** [comment]
@driesvints 
I think the only way to workaround is to use a fallback route now, then I found this piece of code, which I don't know what's the point of this.

https://github.com/laravel/framework/blob/7.x/src/Illuminate/Routing/CompiledRouteCollection.php#L132-L142

The try block always thrown because `$this->routes` always is an empty collection of routes right? If that's true, why will you need such piece of code?

 Could you give me some information about this? Thanks.

**Comment by yaquawa** [comment]
@driesvints By the way, do you know if there is a way to use my own match strategy against the request after this PR?

**Comment by driesvints** [comment]
> then I found this piece of code, which I don't know what's the point of this.

To handle dynamic route arguments properly. We have tests to back that.

> why will you need such piece of code?

The framework supports dynamically added routes at runtime. This collection holds those routes.

> By the way, do you know any other ways to use my own "match strategy" against the request after this PR?

Not at the moment no.

**Comment by yaquawa** [comment]
@driesvints Thanks! Got it 👍🏼

**Comment by yaquawa** [comment]
@driesvints

> The framework supports dynamically added routes at runtime. This collection holds those routes.

But I think the "dynamic" ones will never get checked through the `CompiledUrlMatcher`, which means it never get matched, anything I've missed?

**Comment by driesvints** [comment]
We have tests that verify exactly that.
